### PR TITLE
[tests] Add DeepSeek V4 integration smoke

### PR DESCRIPTION
**Pull request by YeonwooSung**
## Why

DeepSeek V4 (`#3634`) already has debug recipes and CPU unit tests, but it does not appear in `tests/integration_tests/models.py`. The models suite never exercises V4 sparse attention / mHC / hash routing end to end.

## What

- Add a 4-GPU Fake-PG V4 case: FSDP2 + TP2 + EP2, no CP (V4 `update_from_config` rejects CP). Matches the smoke command in the V4 README. Skip ROCm; no golden.
- Register the name in `test_models_select_fake_and_real_pg_cases` and document it in `tests/README.md`.

An earlier revision also added a Qwen3.8 twin of the existing Qwen3.5 FSDP4+TP2+EP4 case. That is dropped: Qwen3.8 shares the Qwen3.5 architecture, so the extra job did not add coverage.

## Test plan

- [x] AST-parse the edited Python files
- [x] CPU: `pytest tests/unit_tests/cpu/test_integration_test_definitions.py -k fake_and_real`
- [ ] GPU CI models suite Fake-PG: `deepseek_v4_fsdp+tp+ep`

### CODE REVIEW

File: tests/integration_tests/models.py
Comment (tianyu-l) [inline tests/integration_tests/models.py:135]: qwen 3.8 and 3.5 share the same arch, maybe we should modify this instead of adding a new one?

File: tests/unit_tests/cpu/test_integration_test_definitions.py [hunk 1]
```diff
@@ -125,14 +129,32 @@ def test_models_select_fake_and_real_pg_cases() -> None:
...
         "qwen3_moe_fsdp+tp+cp+ep_param_groups",
+        "qwen3_8_moe_fsdp+tp+ep",
         "kimi_k2_5_muon_fsdp+ep",
         "muse_glimmer_text_fsdp",
         "muse_glimmer_mm_fsdp+tp+sp",
     } <= fake_pg_model_tests
     assert {"deepseek_v3_fsdp+cp+pp+ep"} <= real_pg_model_tests
 
 
+def test_v4_and_qwen38_smoke_recipes_match_fake_pg_topology() -> None:
```
Comment (tianyu-l) [inline tests/unit_tests/cpu/test_integration_test_definitions.py]: looks useless

File: tests/integration_tests/models.py
Comment (YeonwooSung) [reply tests/integration_tests/models.py:135]: Agreed. Qwen3.8 reuses the Qwen3.5 architecture and this case was the same FSDP4+TP2+EP4 Fake-PG job. I dropped the extra Qwen3.8 entry instead of cloning or replacing the existing Qwen3.5 one. DeepSeek V4 stays; that path is not covered by Qwen3.5.

File: tests/unit_tests/cpu/test_integration_test_definitions.py (hunk 1 above)
Comment (YeonwooSung) [reply tests/unit_tests/cpu/test_integration_test_definitions.py]: Removed. It only re-read the degrees the recipe just set.

**Review by tianyu-l — APPROVED** [review]

**Comment by tianyu-l** [comment]
could you take a look at the CI failure

**Comment by YeonwooSung** [comment]
The Fake-PG failure is only `deepseek_v4_fsdp+tp+ep`. It dies in `verify_module_protocol()` because V4 always attaches `mtp_layers` as `torch.nn.ModuleList` — even when `n_mtp_layers=0`, so the empty list is still a submodule. I don’t think we should copy V3’s `mtp_layers = None` path. I’ll keep V4’s always-a-list shape and switch the container to `torchtitan.protocols.module.ModuleList`.

**Comment by tianyu-l** [comment]
now a different error

**Comment by YeonwooSung** [comment]
`verify_module_protocol` is past. The new Fake-PG failure is the first MoE hash layer: `scores` is `[256, 4]` but hash ids are `[512, 3]`. With TP2 and default sequence parallel, hidden is sharded on T, while `tid2eid[input_ids_T]` still sees the full sequence. I’d rather not shrink the smoke topology and hide this. I can take a follow-up to align V4 hash-routing shard of `input_ids_T` with `x` — okay to do that here, or as a separate PR?

**Comment by tianyu-l** [comment]
@YeonwooSung thanks for investigation, please do it in this PR

**Comment by YeonwooSung** [comment]
Have fixed the issues in the previous CI. WIll wait for the next CI run :)

**Comment by tianyu-l** [comment]
see results

**Comment by YeonwooSung** [comment]
Looked at the latest Fake-PG failure. This one is past both earlier issues: `verify_module_protocol` passes, the model builds, and the MoE hash routing no longer mismatches (`Training starts at step 1` is reached). It now dies at the step-1 finite check with a non-finite loss/grad_norm.

The `No valid triton configs ... triton_flex_attention Required: 163840 Hardware limit:101376` lines are autotuning noise, not the cause: with V4's `head_dim=256`, the large-block FlexAttention configs exceed the A10G's ~100KB shared memory and get ignored, but the autotuner still finds valid small-block kernels (fwd `triton_flex_attention_6`, bwd `triton_flex_attention_backward_15`) and attention runs. The real failure is a genuine non-finite grad_norm from the backward.

This branch was based on an older `main` (390e2985b), predating a few relevant fixes, most notably #4484 ([router] Fp32 for backward), which keeps the MoE router gate in fp32 through backward instead of only fp32 forward autocast. Since the failure is specifically a non-finite grad_norm (backward), that is the most likely fix. I rebased onto current `main` (also picks up #4513, #4498, #4497; goldens were bumped in #4484 so the old base was numerically stale) and re-ran CI. The smoke topology (FSDP2+TP2+EP2, seq 512) is unchanged.

If the non-finite persists after the rebase, it points to a V4-specific training-numerics issue this smoke is the first to exercise end-to-end (V3 debug trains fine on the same runner), and I would rather report/track that than shrink the topology to hide it. Will watch the run.

**Comment by YeonwooSung** [comment]
I reproduced the latest Fake-PG failure on a single L4 (g6.2xlarge, sm89, ~100KB smem, torch nightly cu130) — it matches CI exactly (rc=134, `Loss or gradient norm is not finite at step 1`). Full findings:

**It is not a NaN and not a shape/sharding crash — it is a gradient explosion.** Forward is fine (`loss=7.36`), but `grad_norm=inf`: individual grads are all finite yet reach ~`1e19`, so the fp32 L2 norm overflows to inf. The `No valid triton configs` FlexAttention lines are just A10G/L4 smem autotuning noise — attention compiles and runs on smaller blocks.

**Bisected across parallelism dims (seq 512, 1-step, `--debug`-free, per-parameter grad norms):**

| config | result | grad |
|---|---|---|
| no parallel | ok | grad_norm 4.15 |
| FSDP2 only | ok | normal |
| EP2 only (tp=1) | ok | normal |
| **TP2 + SP** | **fail** | `tok_embeddings` grad 8.8e19 |
| TP2, `--no-enable-sequence-parallel` | **ok** | grad_norm 2.84 |

So the trigger is **Sequence Parallel**, not TP-head-sharding, EP, or FSDP. This also rules out the hash-routing `input_ids_T` change in this PR: that path only runs under SP+EP, but the explosion happens at TP+SP with `ep=1`.

**Localized with backward hooks:** the amplification is at the RMSNorms — `attention_norm`, `attention.q_norm`, `attention.kv_norm` each multiply the incoming gradient by ~1000 per layer (`gout 1.66e16 -> gin 1.66e19`), compounding across the 4 layers up to the embedding. The factor is exactly `1/sqrt(norm_eps)` (`eps=1e-6 -> 1000`), i.e. the norm backward behaves as if its saved input had ~zero variance, even though the forward input is normal (`in_absmean 1.6`, `out 0.8`) and forward loss is correct. Tensors are plain-local under the `spmd_types` backend, no in-place is flagged by anomaly mode, and it reproduces with a hand-written RMSNorm too.

**Conclusion:** this is a pre-existing DeepSeek V4 + Sequence-Parallel gradient bug that this smoke is the first to exercise end-to-end (no V3/V4 integration test currently runs TP+SP, and 8-GPU real-PG has no V4). It is independent of the test scaffolding and of the hash+SP change here.

I'd like to fix it in this PR as you suggested. Before I settle on the fix, one question on direction: do you want the V4 SP gradient path itself corrected (this looks like it lives in the V4 attention/HC sharding, and I'm happy to dig further with the maintainers of that code), or would you prefer the smoke to land now covering FSDP2+TP2+EP2 with SP off (tested green, grad_norm 2.84) plus a tracking issue for the SP gradient bug? I did not shrink the parallelism topology to hide this.

**Comment by YeonwooSung** [comment]
Reproduced the failure on a single L4 and debugged it down to root cause, then moved the case to a real PG. The path:

1. **Reproduced** `deepseek_v4_fsdp+tp+ep` under Fake PG on one L4 (torch nightly cu130): same `rc=134`, `Loss or gradient norm is not finite at step 1`.
2. **Not a NaN, a grad explosion.** Forward is fine (`loss ~7.4`); grads are all finite but reach ~`1e19`, so the fp32 grad-norm overflows to inf. The `No valid triton configs` FlexAttention lines are just A10G/L4 smem autotuning noise (it still compiles on smaller blocks).
3. **Bisected the parallelism:** no-parallel / FSDP-only / EP-only / TP-with-SP-off all train fine (grad_norm ~2.8-4.2); only **TP + SP** explodes. This also rules out the hash-routing `input_ids_T` change here, since that path only runs under SP+EP while the blow-up happens at TP+SP with `ep=1`.
4. **Confirmed it is Fake-PG-specific.** Ran the exact same FSDP2+TP2+EP2 config on a **real 4-GPU PG** (torchrun, 4x L4): it trains cleanly for 10 steps (`loss 8.0 -> 3.7`, `grad_norm ~2-4`). So the model/sharding is correct; under Fake PG the sequence-parallel collectives return activations that alias their inputs, which corrupts a saved-for-backward tensor and drives the explosion.

Given that, I marked the case `use_real_pg=True` so it runs in the 8-GPU real-PG suite and is skipped by the 1-GPU Fake PG job — the same treatment as `varlen_attn+per_op_sac` (#4149), which is kept on a real PG for an analogous Fake-PG/spmd_types incompatibility. Verified on 4x L4: the real-PG harness case passes and the Fake-PG selection now skips it; the selection unit test and README are updated to match.

File: tests/integration_tests/models.py [hunk 2]
```diff
@@ -86,6 +86,22 @@ def build_model_tests_list() -> list[OverrideDefinitions]:
...
         ),
+        OverrideDefinitions(
+            configs=[recipes.deepseek_v4_debugmodel_fsdp2_tp2_ep2],
+            test_descr="DeepSeek V4 FSDP+TP+EP",
+            test_name="deepseek_v4_fsdp+tp+ep",
+            ngpu=4,
+            # Sparse attention / indexer kernels are CUDA-only and unvalidated
+            # on ROCm.
+            skip_rocm_test=True,
+            # Runs on a real PG. Under Fake PG this config's sequence-parallel
```
Comment (wwwjn) [inline tests/integration_tests/models.py:97]: This fake-pg behaviors is not specific to DSV4, but all the models. If you check the `tests/assets/losses/fake_pg`, some of the model gradient norm also explodes. 

The purpose of fake_pg is quickly guard if the composbilit

**Review by wwwjn — CHANGES_REQUESTED** [review]

File: tests/integration_tests/__init__.py [hunk 3]
```diff
@@ -77,6 +77,12 @@ def validate_fake_pg_compatibility(
     # that interaction is fixed. Issue #4149.
     if "varlen_attn+per_op_sac" in test.test_name:
         incompatibilities.append("FSDP + selective AC under spmd_types")
+    # TODO: DeepSeek V4 sequence-parallel collectives return activations that
```
Comment (wwwjn) [inline tests/integration_tests/__init__.py]: Don't add it here.  For this deepseek-v4 testcase, you can still run it with ` use_real_pg=False` without setting any gold_loss file path, so the fake-pg can still guard and find early issue, without worry about the loss explosion behavior.

varlen_attn+per_op_sac (https://github.com/pytorch/torchtitan/issues/4149) this test is not runnable with Fake-pg, but for deepseek-v4, the model is runnable with fake-pg but the loss explodes.

File: tests/integration_tests/models.py (hunk 2 above)
Comment (YeonwooSung) [reply tests/integration_tests/models.py:97]: Agreed that Fake-PG grad-norm blowups are normal and captured in the goldens - qwen3 hits ~2.98e17 (3.8e13 at step 1) and llama3 ~2.36e6. The difference here is that those stay finite, while deepseek_v4's grad_norm goes non-finite (inf). The loss itself is fine (~7.4); it's grad_norm that overflows fp32 to inf, which trips the step-finiteness assert added in #4226 (`torch._assert_async(step_is_finite)` in trainer.py) and aborts at step 1 with rc=134. That is exactly the state this case was already in (use_real_pg=False, no golden), and it is what failed the last few Fake-PG CI runs. On a real 4-GPU PG the same FSDP2+TP2+EP2 config trains cleanly (10 steps, loss 8.0 -> 3.7, grad_norm ~2-4).

File: tests/integration_tests/__init__.py (hunk 3 above)
Comment (YeonwooSung) [reply tests/integration_tests/__init__.py]: You are right that this is semantically different from varlen_attn+per_op_sac (#4149), which is not runnable under Fake PG at all - v4 does build and run. I can drop this addition to validate_fake_pg_compatibility.

The catch for keeping it in the Fake-PG lane is the non-finite grad_norm from the other thread: with use_real_pg=False and no golden it still aborts at step 1 (rc=134) through the #4226 finiteness assert, because v4's grad_norm is inf rather than large-but-finite like qwen3/llama3. So Fake PG cannot currently guard this case - it never completes a step.

Would you prefer I (a) keep it on a real PG (verified: 10 steps, grad_norm ~2-4), or (b) look for a way to keep the Fake-PG grad finite so it can stay in that lane? Happy to go either way.

**Comment by Serendipity-matong** [comment]
I prepared the MTP SPMD layout fix as a follow-up to this PR.
The baseline failed all four MTP layout checks on both ranks:
- mtp_input_embed
- prev_hc_hidden
- mtp_input_ids_T
- mtp_input_valid_mask
After the fix, both ranks passed all checks. The unit test also passed:
2 passed in 7.41s
The follow-up commits are:
- 9a98e6923 [DeepSeek V4] Fix MTP SPMD input layouts
- 6c742b69c [DeepSeek V4] Apply HC rank layouts consistently
They are available on my branch:
https://github.com/Serendipity-matong/torchtitan/tree/followup/4490-deepseek-v4-mtp-spmd-layout
Could you cherry-pick these commits into #4490? Once included, #4521 can be closed as superseded.

**Comment by YeonwooSung** [comment]
Thanks for the fix! I looked at both commits. They target the MTP SPMD layouts, which apply to `deepseek_v4_mtp_debugmodel` (n_mtp_layers=1). This PR's smoke (`deepseek_v4_fsdp+tp+ep`) uses `deepseek_v4_debugmodel` with n_mtp_layers=0, so `config.mtp_layers is None` and that sharding block is skipped; the non-MTP changes only touch the non-SP branch while the smoke runs TP+SP. So these don't affect this PR's CI failure. I'd keep them in their own PR (#4521) rather than cherry-picking into #4490 -- happy to review there.

File: tests/integration_tests/__init__.py (hunk 3 above)
Comment (wwwjn) [reply tests/integration_tests/__init__.py]: Thanks for details! Given #4226 , I'd prefer to do (a) which is simpler. I approved the PR.

My previous experience with (b) is it's quiet hard to make the fake-pg grad finite without changing fake-pg behavior, and it's hard to find a new behavior of fake-pg which is reasonable

**Review by wwwjn — APPROVED** [review]
